File: examples/bench/collatz_compare.py
#!/usr/bin/env python3


def collatz_steps(n: int) -> int:
    steps = 0
    while n != 1:
        if (n % 2) == 0:
            n //= 2
        else:
            n = 3 * n + 1
        steps += 1
    return steps


def sum_collatz(n: int) -> int:
    acc = 0
    while n > 0:
        acc += collatz_steps(n)
        n -= 1
    return acc


if __name__ == "__main__":
    print(sum_collatz(15000))

File: examples/bench/collatz_compare.f
# Branch-heavy integer recursion benchmark.

collatz_steps(1): 0.
collatz_steps(n | n % 2 = 0): 1 + collatz_steps(n / 2).
collatz_steps(n): 1 + collatz_steps(3 * n + 1).

sum_collatz(0, acc): acc.
sum_collatz(n, acc): sum_collatz(n - 1, acc + collatz_steps(n)).

main():
    n <- 15000
    say(sum_collatz(n, 0))
    1.
